discovery/src/main/java/org/wildfly/extension/discovery/StaticProviderAddHandler.java

```diff
@@ -0,0 +1,116 @@
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.extension.discovery;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.jboss.as.controller.AbstractAddStepHandler;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.as.controller.operations.global.WriteAttributeHandler;
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.dmr.ModelNode;
+import org.wildfly.discovery.AttributeValue;
+import org.wildfly.discovery.ServiceURL;
+import org.wildfly.discovery.impl.MutableDiscoveryProvider;
+import org.wildfly.discovery.impl.StaticDiscoveryProvider;
+import org.wildfly.discovery.spi.DiscoveryProvider;
+
+/**
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ */
+class StaticProviderAddHandler extends AbstractAddStepHandler {
+    private static final StaticProviderAddHandler INSTANCE = new StaticProviderAddHandler();
+
+    static StaticProviderAddHandler getInstance() {
+        return INSTANCE;
+    }
+
+    private StaticProviderAddHandler() {
+        super(new Parameters().addAttribute(StaticProviderDefinition.SERVICES));
+    }
+
+    public void execute(final OperationContext context, final ModelNode operation) throws OperationFailedException {
+        context.registerCapability(
+            RuntimeCapability.Builder
+                .of(DiscoveryExtension.DISCOVERY_PROVIDER_CAPABILITY, true, new MutableDiscoveryProvider())
+                .build().fromBaseCapability(context.getCurrentAddressValue()));
+        super.execute(context, operation);
+    }
+
+    protected void recordCapabilitiesAndRequirements(final OperationContext context, final ModelNode operation, final Resource resource) throws OperationFailedException {
+        // no operation
+    }
+
+    protected void performRuntime(final OperationContext context, final ModelNode operation, final Resource resource) throws OperationFailedException {
+        modifyRegistrationModel(context, operation);
```

Comment:
WriteAttributeHandler.INSTANCE is a Stage.MODEL handler and will fail if called from performRuntime, which executes in Stage.RUNTIME. Also, WriteAttributeHandler won't understand the 'operation' ModelNode passed in from performRuntime as an 'add' operation has a different structure of parameters from a 'write-attribute'.